modules/placement-driver/src/main/java/org/apache/ignite/internal/placementdriver/LeaseUpdater.java

```diff
@@ -0,0 +1,242 @@
+    }
+
+    /**
+     * Activates a lease updater to renew leases.
+     */
+    public void activate() {
+        //TODO: IGNITE-18879 Implement lease maintenance.
+        updaterTread = new Thread(updater, NamedThreadFactory.threadPrefix(nodeName, "lease-updater"));
+
+        updaterTread.start();
+    }
+
+    /**
+     * Runnable to update lease in Metastorage.
+     */
+    private class Updater implements Runnable {
+        @Override
+        public void run() {
+            while (!updaterTread.isInterrupted()) {
+                for (Map.Entry<ReplicationGroupId, Set<Assignment>> entry : assignmentsTracker.assignments().entrySet()) {
+                    ReplicationGroupId grpId = entry.getKey();
+
+                    Lease lease = leaseTracker.getLease(grpId);
+
+                    HybridTimestamp now = clock.now();
+
+                    // Nothing holds the lease.
+                    if (lease.getLeaseExpirationTime() == null
+                            // The lease is near to expiration.
+                            || now.getPhysical() > (lease.getLeaseExpirationTime().getPhysical() - LEASE_PERIOD / 2)) {
+                        var leaseKey = ByteArray.fromString(PLACEMENTDRIVER_PREFIX + grpId);
+
+                        var newTs = new HybridTimestamp(now.getPhysical() + LEASE_PERIOD, 0);
+
+                        ClusterNode candidate = nextLeaseHolder(entry.getValue());
+
+                        // Now nothing is holding the lease, and new lease is being granted.
+                        if (lease.getLeaseholder() == null && candidate != null
+                                // The lease is prolongation
+                                || lease.getLeaseholder() != null && lease.getLeaseholder().equals(candidate)
+                                // New lease is being granted, and the previous lease is expired.
+                                || candidate != null && (lease.getLeaseExpirationTime() == null
+                                || compareWithClockSkew(now, lease.getLeaseExpirationTime()) > 0)) {
+                            byte[] leaseRaw = ByteUtils.toBytes(lease);
+
+                            Lease renewedLease = new Lease(candidate, newTs);
+
+                            msManager.invoke(
+                                    or(notExists(leaseKey), value(leaseKey).eq(leaseRaw)),
+                                    put(leaseKey, ByteUtils.toBytes(renewedLease)),
+                                    noop()
+                            );
+                        }
+                    }
+                }
+
+                try {
+                    Thread.sleep(UPDATE_LEASE_MS);
+                } catch (InterruptedException e) {
+                    LOG.warn("Lease updater is interrupted");
+                }
+            }
+        }
+    }
+
+    /**
+     * Compares two timestamps with the clock skew.
+     * TODO: IGNITE-18978 Method to comparison timestamps with clock skew.
+     *
```

Comment:
Javadoc should describe what exactly this method does.

Reply:
Is it not right?